x/config/spec_test.go

```diff
@@ -0,0 +1,643 @@
+			compiler: compileUnaryOutboundConfig,
+			attrs:    attributeMap{"key": "key", "value": "value"},
+			want:     someItem,
+		},
+		{
+			desc:     "decode *item",
+			build:    func(*item, transport.Transport) (transport.UnaryOutbound, error) { panic("kthxbye") },
+			compiler: compileUnaryOutboundConfig,
+			attrs:    attributeMap{"key": "key", "value": "value"},
+			want:     ptrToSomeItem,
+		},
+		{
+			desc:     "decode **item",
+			build:    func(**item, transport.Transport) (transport.UnaryOutbound, error) { panic("kthxbye") },
+			compiler: compileUnaryOutboundConfig,
+			attrs:    attributeMap{"key": "key", "value": "value"},
+			want:     &ptrToSomeItem,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			spec, err := tt.compiler(tt.build)
+			if !assert.NoError(t, err, "failed to compile config") {
+				return
+			}
+
+			got, err := spec.Decode(tt.attrs)
+			if len(tt.wantErr) == 0 {
+				if assert.NoError(t, err) {
+					assert.Equal(t, tt.want, got.data.Interface())
+				}
+				return
+			}
+
+			if assert.Error(t, err, "expected failure") {
+				for _, msg := range tt.wantErr {
+					assert.Contains(t, err.Error(), msg)
+				}
+			}
+		})
+	}
+}
+
+// mockValueBuilder is a simple callable that records and verifies its calls using
+// a gomock controller.
+//
+// m.Build is a valid builder function for configuredValue for some
+// mockValueBuilder m.
+type mockValueBuilder struct{ ctrl *gomock.Controller }
+
+func newMockValueBuilder(ctrl *gomock.Controller) *mockValueBuilder {
+	return &mockValueBuilder{ctrl: ctrl}
+}
+
+func (m *mockValueBuilder) ExpectBuild(args ...interface{}) *gomock.Call {
+	return m.ctrl.RecordCall(m, "Build", args...)
+}
+
+func (m *mockValueBuilder) Build(args ...interface{}) (interface{}, error) {
+	ret := m.ctrl.Call(m, "Build", args...)
+	err, _ := ret[1].(error)
+	return ret[0], err
+}
+
+func TestConfiguredValueDecode(t *testing.T) {
+	type item struct{ Key, Value string }
+
+	tests := []struct {
+		desc string
+
+		// Configuration data and arguments for the build function
+		data interface{}
+		args []interface{}
+
+		// Expect a Build(..) call with the given arguments
+		wantArgs []interface{}
+
+		// Result and error of calling the build function
+		result interface{}
+		err    error
+	}{
+		{
+			desc:     "success, no args",
+			data:     struct{}{},
+			wantArgs: []interface{}{struct{}{}},
+			result:   42,
+		},
+		{
+			desc:     "success with args",
+			data:     1,
+			args:     []interface{}{2, 3},
+			wantArgs: []interface{}{1, 2, 3},
+			result:   4,
+		},
+		{
+			desc: "success with Value args",
+			data: &item{Key: "key", Value: "value"},
+			args: []interface{}{
+				"foo",
+				reflect.ValueOf("bar"),
+				"baz",
+			},
+			wantArgs: []interface{}{
+				&item{Key: "key", Value: "value"},
+				"foo",
+				"bar",
+				"baz",
+			},
+			result: `¯\_(ツ)_/¯`,
```

Comment:
should we just remove the result since we're not really comparing it against anything, we could get the same test coverage by generating a random string